server_addon/houdini/client/ayon_houdini/plugins/publish/collect_files_for_cleaning_up.py

```diff
@@ -0,0 +1,112 @@
+import os
+
+import clique
+import pyblish.api
+from ayon_core.pipeline import AYONPyblishPluginMixin
+from ayon_houdini.api import plugin
+
+
+class CollectFilesForCleaningUp(plugin.HoudiniInstancePlugin,
+                                AYONPyblishPluginMixin):
+    """Collect Files For Cleaning Up.
+
+    This collector collects output files
+    and adds them to file remove list.
+
+    CAUTION:
+        This collector registers exported files and
+          the parent folder (if it was empty) for deletion
+          in `ExplicitCleanUp` plug-in.
+        Artists are free to change the file path in the ROP node.
+    """
+
+    # It should run after CollectFrames and Collect Render plugins,
+    # and before CollectLocalRenderInstances.
+    order = pyblish.api.CollectorOrder + 0.115
+
+    hosts = ["houdini"]
+    families = ["*"]
+    label = "Collect Files For Cleaning Up"
+    intermediate_exported_render = False
+
+    def process(self, instance):
+
+        files = []
+        staging_dirs = []
+
+        expected_files = instance.data.get("expectedFiles", [])
+
+        # 'expectedFiles' are preferred over 'frames'
+        if expected_files:
+            # Products with expected files
+            # This can be Render products or submitted cache to farm.
+            for expected in expected_files:
+                # expected.values() is a list of lists
+                for output_files in expected.values():
+                    staging_dir, _ = os.path.split(output_files[0])
+                    if staging_dir not in staging_dirs:
+                        staging_dirs.append(staging_dir)
+                    files.extend(output_files)
+        else:
+            # Products with frames or single file.
+
+            staging_dir = instance.data.get("stagingDir")
+            staging_dirs.append(staging_dir)
+
+            frames = instance.data.get("frames")
+            if frames is None:
+                self.log.warning(
+                    f"No frames data found on instance {instance}"
+                    ". Skipping collection for caching on farm..."
+                )
+                return
+            elif isinstance(frames, str):
+                # single file.
+                files.append(frames)
+            else:
+                # list of frame.
+                files.extend(
+                    [f"{staging_dir}/{frame}" for frame in frames]
+                )
```

Comment:
We don't ever want `None` in here I suppose? So this `instance.data.get("stagingDir")` maybe needs to be required? 

Also I think you're lacking making the `files` complete for the single file condition?

```suggestion
            frames = instance.data.get("frames")
            if frames is None:
                self.log.warning(
                    f"No frames data found on instance {instance}"
                    ". Skipping collection for caching on farm..."
                )
                return

            staging_dir = instance.data.get("stagingDir")
            staging_dirs.append(staging_dir)

            if isinstance(frames, str):
                # single file.
                files.append(f"{staging_dir}/{frames}")
            else:
                # list of frame.
                files.extend(
                    [f"{staging_dir}/{frame}" for frame in frames]
                )
```